Sources/SwiftProtobuf/TextDecoder.swift

```diff
@@ -0,0 +1,211 @@
+        }
+        if terminator == nil {
+            return
+        } else {
+            throw DecodingError.truncatedInput
+        }
+    }
+}
+
+struct TextFieldDecoder: FieldDecoder {
+    var rejectConflictingOneof: Bool {return true}
+    var scanner: TextScanner
+
+    mutating func decodeExtensionField(values: inout ExtensionFieldValueSet, messageType: Message.Type, protoFieldNumber: Int) throws {
+        if let ext = scanner.extensions?[messageType, protoFieldNumber] {
+            var mutableSetter: FieldDecoder = self
+            var fieldValue = values[protoFieldNumber] ?? ext.newField()
+            try fieldValue.decodeField(setter: &mutableSetter)
+            values[protoFieldNumber] = fieldValue
+        }
+    }
+
+    mutating func decodeSingularField<S: FieldType>(fieldType: S.Type, value: inout S.BaseType?) throws {
+        try scanner.skipRequired(token: .colon)
+        try S.setFromText(scanner: scanner, value: &value)
+    }
+
+    mutating func decodeRepeatedField<S: FieldType>(fieldType: S.Type, value: inout [S.BaseType]) throws {
+        try scanner.skipRequired(token: .colon)
+        if try scanner.skipOptional(token: .beginArray) {
+            var firstItem = true
+            while true {
+                if try scanner.skipOptional(token: .endArray) {
+                    return
+                }
+                if firstItem {
+                    firstItem = false
+                } else {
+                    try scanner.skipRequired(token: .comma)
+                }
+                try S.setFromText(scanner: scanner, value: &value)
+            }
+        } else {
+            try S.setFromText(scanner: scanner, value: &value)
+        }
+    }
+
+    mutating func decodePackedField<S: FieldType>(fieldType: S.Type, value: inout [S.BaseType]) throws {
+        try decodeRepeatedField(fieldType: fieldType, value: &value)
+    }
+
+    mutating func decodeSingularMessageField<M: Message>(fieldType: M.Type, value: inout M?) throws {
+        _ = try scanner.skipOptional(token: .colon)
+        try M.setFromText(scanner: scanner, value: &value)
+    }
+
+    mutating func decodeRepeatedMessageField<M: Message>(fieldType: M.Type, value: inout [M]) throws {
+        _ = try scanner.skipOptional(token: .colon)
+        if try scanner.skipOptional(token: .beginArray) {
+            var firstItem = true
+            while true {
+                if try scanner.skipOptional(token: .endArray) {
+                    return
+                }
+                if firstItem {
+                    firstItem = false
+                } else {
+                    try scanner.skipRequired(token: .comma)
+                }
+                try M.setFromText(scanner: scanner, value: &value)
+            }
+        } else {
+            try M.setFromText(scanner: scanner, value: &value)
+        }
+    }
+
+    mutating func decodeSingularGroupField<G: Message>(fieldType: G.Type, value: inout G?) throws {
```

Comment:
Again, this is called by the generated code and communicates the declared type of the field to the decoder.  For Text and JSON formats, there's no difference, but Protobuf Binary does distinguish groups and messages on the wire.